Adding plugin to read and write to Universal Spectroscopy and Imaging Data HDF5 files

ssomnath opened the pull request:
### Description of the change
- Added new io plugin for USID
- Added thorough unit tests

### Progress of the PR
- [ ] Change implemented (can be split into several points),
- [x] no need to update docstrings
- [x] already updated user guide,
- [x] already added tests,
- [x] ready for review.

### Minimal example of the bug fix or the new feature
```python
>>> import hyperspy.api as hs
>>> import numpy as np
>>> s = hs.signals.Signal1D(np.arange(10))
>>> # Save signal to USID compliant HDF5 file:
>>> s.save('USID.h5')
>>> # Load signal(s) from USID HDF5 files
>>> new_signal = hs.load('USID.h5')
```

ssomnath committed:
First version of USID reader and writer

ssomnath committed:
Added pyUSID as an optional requirement

ssomnath committed:
Added the USID reader to the list of io plugins

ssomnath committed:
Renamed module to avoid potential namespace conflicts

ssomnath committed:
Added docstrings to helper functions

ssomnath committed:
Fixed bugs that surfaced during unit testing

ssomnath committed:
Working tests for USID io plugin

ssomnath committed:
Minor PEP8 cleaning

ssomnath committed:
Cleaned up imports

ssomnath committed:
Test only if installed

ssomnath committed:
Test only if installed

ssomnath committed:
Tested to ensure that tests don't fail if pyUSID is unavailable

ssomnath committed:
Shortened the name of a keyword argument for file_reader
Reading from h5USID files default to ignoring non-linear dimensions
with warnings instead of throwing errors
More descriptive warnings and error messages that point to the
appropriate keyword argument to solve a problem

ssomnath committed:
Added pyUSID to Travis testing dependencies

ssomnath committed:
Added comprehensive documentation for using the new io plugin

ericpre commented on doc/user_guide/io.rst:
```diff
@@ -212,6 +212,8 @@ HyperSpy. The "lazy" column specifies if lazy evaluation is supported.
     +--------------------+--------+--------+--------+
     | EDAX .spc and .spd |    Yes |    No  |    Yes |
     +--------------------+--------+--------+--------+
+    | h5USID .h5         |    Yes |    No  |    No  |
```
It seems that it can write too!

ssomnath replied:
> * At some point (fairly soon?), support for non linear axis will be added in hyperspy (#1927), it would be nice if you could add some comments to what would be necessary to change when the non linear axis branch is merged.

OK. I added notes in _get_dim_dict()

> * it would be useful to support lazy loading with dask.

Indeed. This is being tested and will be pushed in the next release of pyUSID soon.

> * out of curiosity, why using two underscore `__` instead of only one `_` which seems to be the convention in python.

double underscore prevents those objects from being used outside the module in which they are declared. I typically use these for functions which generally have no purpose outside the module. I typically use single underscore for those internal functions that may be useful for those who know what they are doing. 

> * Limit all lines to a maximum of 79 characters, https://www.python.org/dev/peps/pep-0008/#maximum-line-length

OK. Done.

ssomnath replied:
> It seems that it can write too!

Thanks for catching this typo. I fixed it

ssomnath replied:
Lazy loading and saving now supported

ericpre commented on hyperspy/io_plugins/usid_hdf5.py:
```diff
@@ -0,0 +1,447 @@
+                              'Please delete the file at the location or specify a '
+                              'different path for the file'.format(filename))
+    """
+    # CANNOT import BaseSignal! Probably circular import
```
Add it here, in this function and it should work fine.
```python
from hyperspy.signal import BaseSignal
```

ssomnath replied:
Turns out that I don't need to import the BaseSignal class after all so this is fine for now I guess. One question I have is how do I return a dataset (as a dictionary) which has N navigation axes and M signal axes? At the moment, this information is lost in the translation. When I [originally developed an ingestor](https://gist.github.com/ssomnath/097d2bb7c1468709ce2899512deb5b6d) - see lines 97 - 124, I was able to move the signal axes... I am not sure how to express this in the dictionary that I return now

ericpre commented on hyperspy/io_plugins/usid_hdf5.py:
```diff
@@ -0,0 +1,447 @@
+    """
+
+    # Not sure how to safely ignore spurious / additional dimensions
+    if len(object2save.axes_manager.shape) != len(object2save.data.shape):
```
It should not be necessary to check this and this will actually not work with "ragged" array. That being said I don't think that there is any format in hyperspy supporting ragged array without workaround.

ericpre commented on setup.py:
```diff
@@ -73,6 +73,7 @@
     "gui-traitsui": ["hyperspy_gui_traitsui>=1.1.0"],
     "mrcz": ["blosc>=1.5", 'mrcz>=0.3.6'],
     "speed": ["numba"],
+    "usid": ["pyUSID"],
```
I should be fine to add `pyUSID` as required dependency since it is available from the conda-forge channel and on pypi. The only major drawback is that it will increase our numpy version requirement to >1.13.

ssomnath replied:
We use only one feature of Numpy 1.13 (can't remember which one) which is why we had to bump up our minimum version number. I can see how we can work around this issue. 

I currently listed pyUSID only as an optional requirement. Did you mean to include it as a core requirement? Is there a reason why bumping numpy requirement to 1.13 would be a major drawback / show-stopper?

ssomnath replied:
The latest version of pyUSID (0.0.6.1) now only needs numpy 1.10. This should alleviate your concern

ericpre commented on doc/user_guide/io.rst:
```diff
@@ -881,6 +883,126 @@ good approximation. Further release of HyperSpy will read the time axis more
+    ValueError: Cannot load provided dataset. Parameter: Bias was varied non-linearly.
+    Supply keyword argument "ignore_non_linear_dims=True" to ignore this error
+
+Lazy reads will be made available in an upcoming version of pyUSID and HyperSpy.
```
What is the reason not to support lazy processing?

ericpre commented on hyperspy/io_plugins/usid_hdf5.py:
```diff
@@ -0,0 +1,447 @@
+    sig = {'data': ndim_form,
+           'axes': dim_dict_list,
+           'metadata': {
+               'Signal':{'quantity': quantity, 'units': units, 'signal_type': sig_type},
```
`units` is not a metadata as defined in the metadata structure: http://hyperspy.org/hyperspy-doc/current/user_guide/metadata_structure.html

ssomnath replied:
OK. I moved all non-standard metadata to original_metadata

ericpre commented on hyperspy/io_plugins/usid_hdf5.py:
```diff
@@ -0,0 +1,447 @@
+           'metadata': {
+               'Signal':{'quantity': quantity, 'units': units, 'signal_type': sig_type},
+               'General': {'original_filename': h5_path, 'dataset_path': h5_dset_path,
+                           'original_file_type': 'USID HDF5', 'pyUSID_version': usid.__version__}
```
same as above for `original_file_type`.

ericpre commented on hyperspy/io_plugins/usid_hdf5.py:
```diff
@@ -0,0 +1,447 @@
+
+    num_spec_dims = len(spec_dict)
+    num_pos_dims = len(pos_dict)
+    if verbose:
```
Use the logger instead of the `print` function?

ssomnath replied:
Done

ericpre commented on hyperspy/io_plugins/usid_hdf5.py:
```diff
@@ -0,0 +1,447 @@
+# ######################################################################################################################
+
+
+def file_reader(filename, dset_path='', verbose=False, ignore_non_linear_dims=True, **kwds):
```
Set `dset_path` default to None? Usually, we load most of the data as a list of signals.

Add parameters specific to this reader to the `load` function: https://github.com/hyperspy/hyperspy/blob/145003e38598d8244625390bc3de368de7db985d/hyperspy/io.py#L46

ssomnath replied:
I don't think it would be a good idea to load dozens of datasets into memory (numpy). We can change the default value to None once lazy reading is implemented. Until then, I would recommend keeping the default to reading a single dataset.

ssomnath replied:
Now, ``dset_path=None`` is the default given that lazy loading is allowed.

ericpre reviewed (commented):
A few very quick comments for now:
- At some point (fairly soon?), support for non linear axis will be added in hyperspy (#1927), it would be nice if you could add some comments to what would be necessary to change when the non linear axis branch is merged.
- it would be useful to support lazy loading with dask.

Formatting:
- out of curiosity, why using two underscore `__` instead of only one `_` which seems to be the convention in python.
- Limit all lines to a maximum of 79 characters, https://www.python.org/dev/peps/pep-0008/#maximum-line-length

ssomnath committed:
Corrected write capability in table
Expanded statement on no lazy read / write (yet).

ssomnath committed:
Replaced all print statements with logger info
PEP8 clean-up - limited line length to 79 characters
Removed verbose kwarg
Moved USID specific metadata to original metadata

ssomnath committed:
PEP8 clean-up - limited line length to 79 characters
Using two new helper functions to reuse code better

ssomnath committed:
Removed unnecessary code

ssomnath committed:
Added instructions on accepting non-linear dimensions into Axes

ssomnath commented:
I think I have addressed nearly all your comments in my fork. Let me know if you can think of any other changes.

ssomnath committed:
Added ability to append to existing HDF5 file

ssomnath committed:
Added test for appending to existing file

ssomnath committed:
Passing on keywords (compression, chunking, etc.) to pyUSID

ssomnath committed:
Added notes about appending to h5USID files

ssomnath committed:
Made changes to allow lazy reading. Lazy writing already supported

ssomnath committed:
Using new get_slope function from pyUSID
Embedding name of dataset as the title
Adding channel and measurement level attributes under extra_metadata
Simplified parsing of descriptor strings

ssomnath committed:
Added tests for lazy loading and saving of signals

ssomnath committed:
Added tests for attributes in parent groups

ssomnath committed:
Stating that lazy read and write capabilities are now available

ssomnath committed:
Correcting minimum version of pyUSID required

ssomnath committed:
Removed option of loading only the first dataset. Load all by default

ssomnath committed:
Simplified documentation for loading.
Added note regarding the need and consequence of lazy=True

ericpre commented:
@ssomnath it would more convenient to keep this PR open and not to open another one in order to keep the discussion in the same place and not have it scattered over several PRs. For the review, it is also more useful when comparing changes. If necessary, "Force-push" to your branch is fine.
Can you please reopen this PR?

ericpre commented on hyperspy/io_plugins/usid_hdf5.py:
```diff
@@ -0,0 +1,521 @@
+        h5_f.close()
+
+
+def file_writer(filename, object2save, *args, **kwds):
```
Remove `*args` since it is not used.

ericpre committed:
Merge remote-tracking branch 'upstream/RELEASE_next_minor' into RELEASE_next_minor-ssomnath

# Conflicts:
#	.travis.yml
#	setup.py

ericpre committed:
Add pyUSID dependency to appveyor.

ericpre commented:
@ssomnath: sorry to coming back very late to this...
I have solve the merge conflict and it seems that recent changes in pyUSID make the USID tests fail. pyUSID 0.0.6.3 and master fails but 0.0.6.2 work fine (it seems that it started from https://github.com/pycroscopy/pyUSID/commit/8ef95004345f1fbb4636b795283d967d0d4e383b).

What the changes between pyUSID 0.0.6.2 and 0.0.6.3, which would make the USID tests of this PR fail?
What would be better to do? Update this PR and use pyUSID>=0.0.6.3 or fix pyUSID and exclude pyUSID 0.0.6.3?

ssomnath committed:
Fixed problem with reversed dimension ordering in saving to h5USUD

ssomnath committed:
Now generating data always from slow to fast varying dims
aux dims can be varied in order though

ssomnath committed:
Fixing the N pos M spec test by flipping the order of the spec dims

ssomnath committed:
Now generating complex and compound valued data within same func

ssomnath committed:
generator func's mode now should be a string to avoid accidents

ssomnath committed:
Calling numpy.random.rand correctly

ssomnath committed:
Explicitly stating the slow to fast mode for the multi-dset tests

ssomnath committed:
Using ArrayTranslator instead of deprecated NumpyTranslator

ssomnath committed:
FileWriter for USID set to fast to slow dimension order.
Needs fixing

ssomnath committed:
FileWriter for USID set to slow to fast dimension order - future correct

ssomnath committed:
Fixed test for non linear dimension by forcing dims -> slow to fast

ssomnath committed:
Adapted test for compound dataset to use base utils from float dataset

ssomnath committed:
Fixed error with complex valued test slow_to_fast=True now

ssomnath committed:
Fixed error with compound valued test slow_to_fast=True now

ssomnath committed:
Updated requirements for pyUSID

ssomnath committed:
Merge branch 'RELEASE_next_minor' into RELEASE_next_minor

ssomnath commented:
@ericpre - bug fixes completed. Would you mind revisiting this again?

ericpre commented on hyperspy/io_plugins/usid_hdf5.py:
```diff
@@ -0,0 +1,519 @@
+    items = []
+    for k, v in nested_dict.items():
+        new_key = parent_key + sep + k if parent_key else k
+        if isinstance(v, collections.MutableMapping):
```
Use `collections.abc.MutableMapping` to avoid a warning.

ericpre commented on doc/user_guide/io.rst:
```diff
@@ -904,6 +906,127 @@ good approximation. Further release of HyperSpy will read the time axis more
+.. code-block:: python
+
+    >>> # Loading a specific dataset
+    >>> hs.load("sample.h5", dset_path='/Measurement_004/Channel_003')
```
Add `dset_path` to the docstring of [`load`](https://github.com/hyperspy/hyperspy/blob/1b32a1bbe82dc089e1517619fe7b53128a1c1e21/hyperspy/io.py#L56)?

ericpre reviewed (commented):
Thanks @ssomnath for the coming back quickly on this!
A few more things and it should be good to merge:
- see the few comments,
- a deprecation warning with `collections` needs to be fixed - see comment,
- https://github.com/conda-forge/pyusid-feedstock/pull/12 needs to be merged to make CI happy (the tests pass nicely locally with pyusid 0.0.7),
- fix the following userwarning when running the test suite:
```
hyperspy/tests/io/test_usid.py::TestHS2USIDallKnown::test_n_pos_0_spec
hyperspy/tests/io/test_usid.py::TestHS2USIDallKnown::test_0_pos_n_spec
hyperspy/tests/io/test_usid.py::TestHS2USIDallKnown::test_n_pos_m_spec
hyperspy/tests/io/test_usid.py::TestHS2USIDallKnown::test_append_to_existing_file
hyperspy/tests/io/test_usid.py::TestHS2USIDallKnown::test_append_to_existing_file
hyperspy/tests/io/test_usid.py::TestHS2USIDlazy::test_base_nd
hyperspy/tests/io/test_usid.py::TestHS2USIDedgeAxes::test_no_axes
hyperspy/tests/io/test_usid.py::TestHS2USIDedgeAxes::test_incorrect_axes
hyperspy/tests/io/test_usid.py::TestUSID2HSbase::test_n_pos_0_spec
hyperspy/tests/io/test_usid.py::TestUSID2HSbase::test_0_pos_n_spec
hyperspy/tests/io/test_usid.py::TestUSID2HSbase::test_n_pos_m_spec
hyperspy/tests/io/test_usid.py::TestUSID2HSbase::test_lazy_load
hyperspy/tests/io/test_usid.py::TestUSID2HSbase::test_lazy_load
hyperspy/tests/io/test_usid.py::TestUSID2HSdtype::test_complex
hyperspy/tests/io/test_usid.py::TestUSID2HSdtype::test_compound
hyperspy/tests/io/test_usid.py::TestUSID2HSdtype::test_non_linear_dimension
hyperspy/tests/io/test_usid.py::TestUSID2HSdtype::test_non_linear_dimension
hyperspy/tests/io/test_usid.py::TestUSID2HSmultiDsets::test_pick_specific
hyperspy/tests/io/test_usid.py::TestUSID2HSmultiDsets::test_read_all_by_default
  /home/eric/Dev/pyUSID/pyUSID/io/usi_data.py:151: UserWarning: This dataset does not have an N-dimensional form
    warn('This dataset does not have an N-dimensional form')

```

ssomnath committed:
Addressing comments from Eric

ssomnath committed:
Added dset_path to docstring for hs.load for USID

ssomnath committed:
Merge remote-tracking branch 'origin/RELEASE_next_minor' into RELEASE_next_minor
